src/ts-types/types.d.ts

```diff
@@ -0,0 +1,134 @@
+export type Mixed = {};
+export type Maybe<T> = T | null | undefined;
+export type Constructor<T = Mixed> = new (...args: any[]) => T;
+export type HTMLElementClass = typeof HTMLElement;
+
+export type ComponentProps<T, E = HTMLElementClass> = {
+  [P in keyof T]: PropOptions
+};
+
+// NOTE:
+// - all classes are just ambient definitions (opaque types like), so consumer cannot use them directly
+// - inferring generics work only on instances, not on implementation type. So this will not give you type safety, you still have to manually annotate those props in your code
+
+/**
+ * Implement this interface for any @skatejs/renderer-*
+ */
+export interface Renderer<O> {
+  // render?(props?: Mixed): O;
+
+  // called after render
+  // rendered?(): void;
+
+  renderer(
+    shadowRoot: Node | Element | ShadowRoot,
+    render: (props?: Mixed) => O
+  ): void;
+}
+
+export interface WithComponent<P = Mixed, S = Mixed, C = Mixed>
```

Comment:
Typescript is flexible, we can extend types via this DRY pattern so there is no need to reimplement  type definitions within `declare class`